pkg/controller/controller_utils.go

```diff
@@ -1073,19 +1073,19 @@ func AddOrUpdateTaintOnNode(ctx context.Context, c clientset.Interface, nodeName
 	if len(taints) == 0 {
 		return nil
 	}
-	firstTry := true
+
 	return clientretry.RetryOnConflict(UpdateTaintBackoff, func() error {
 		var err error
 		var oldNode *v1.Node
-		// First we try getting node from the API server cache, as it's cheaper. If it fails
```

Comment:
Can you explain why the original code can race? 

In the original behavior, In my understanding, the `resourceVersion` in node acts as an a optimistic lock to avoid race. so no matter getting a stale node object in the first try or patching node by another client in the same time, both will leads a conflict error to avoid race.

Reply:
good question, this is my understanding, the code prepares a strategic merge patch `PatchNodeTaints(ctx, c, nodeName, oldNode, newNode)`

https://github.com/kubernetes/kubernetes/blob/0f095cf0ba5156a972b446a5f65c963fe2a7919c/pkg/controller/controller_utils.go#L1169-L1195

It seems the patch will replace all the existing taints

```
        "taints": {
          "description": "If specified, the node's taints.",
          "items": {
            "$ref": "#/definitions/io.k8s.api.core.v1.Taint"
          },
          "type": "array",
          "x-kubernetes-list-type": "atomic"
        },
```

This is better explained here https://kubernetes.io/docs/reference/using-api/api-concepts/#patch-and-apply , AFAIK the resourceVersion applies only to Updates 

cc: @liggitt

Reply:
It's really surprise me, so I just tried in my local, and it seems that `resourceVersion` does work when do `patch` method. 
![image](https://github.com/user-attachments/assets/21e02ae6-6c0d-4a09-9cbe-85c3c71e5dc0)

As the picture above can see, the latest version of the node is `1576187`. When I do patch to modify taint with a stale version `1575874`, an conflict error occureded. And when I do the same thing with the right version `1576187`, patching is successed.

Reply:
If the experiment I did was right, IMO, removing `resourceVersion=0` will not make any difference.

because even if we bypass the cache and get the very latest version of node, it can not be sure that there is no other change between `get` and `patch`.

Reply:
@xuzhenglun I'm not sure that you send `resourceVersion` when you send a PATCH